src/hotspot/cpu/x86/stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,2012 @@
+    const XMMRegister save_rbx  = xmm8;
+    // xmm registers more valuable in inner loops...
+    // const XMMRegister save_rsi  = xmm9;
+    // const XMMRegister save_rdi  = xmm10;
+    // const XMMRegister save_rcx  = xmm11;
+    // const XMMRegister save_r8   = xmm12;
+
+    // Addresses of the two jump tables used for small needle processing
+    address jump_table;
+    address jump_table_1;
+
+    Label L_begin;
+
+    Label L_returnRBP, L_checkRangeAndReturn, L_returnError;
+    Label L_bigCaseFixupAndReturn, L_checkRangeAndReturnRCX;
+    Label L_returnZero, L_copyHaystackToStackDone, L_bigSwitchTop;
+    Label L_bigCaseDefault, L_smallCaseDefault, L_copyHaystackToStack, L_smallSwitchTop;
+    Label L_nextCheck, L_checksPassed, L_zeroCheckFailed;
+
+    Label L_wcharBegin, L_continue, L_wideNoExpand, L_copyDone, L_copyHigh;
+    Label L_wideMidLoop, L_wideTopLoop, L_wideInnerLoop, L_wideFound;
+
+    __ align(CodeEntryAlignment);
+    fnptrs[isLL   ? StrIntrinsicNode::LL
+           : isUU ? StrIntrinsicNode::UU
+                  : StrIntrinsicNode::UL] = __ pc();
+    __ enter();  // required for proper stackwalking of RuntimeStub frame
+
+    // Check for trivial cases
+    // needle length == 0?
+    __ cmpq(needle_len, 0);
+    __ jg_b(L_nextCheck);
+    __ xorq(rax, rax);
+    __ leave();
+    __ ret(0);
+
+    // haystack length == 0?
+    __ bind(L_nextCheck);
+    __ testq(haystack_len, haystack_len);
+    __ je(L_zeroCheckFailed);
+
+    // haystack length >= needle length?
+    __ movq(rax, haystack_len);
+    __ subq(rax, needle_len);
+    __ jge_b(L_checksPassed);
+
+    __ bind(L_zeroCheckFailed);
+    __ movq(rax, -1);
+    __ leave();
+    __ ret(0);
+
+    __ bind(L_checksPassed);
+
+    // Check for highly-optimized ability - haystack <= 32 bytes and needle <= OPT_NEEDLE_SIZE_MAX
+    // haystack_len is in elements, not bytes, for UTF-16
+    __ cmpq(haystack_len, isU ? COPIED_HAYSTACK_STACK_SIZE / 4
+                              : COPIED_HAYSTACK_STACK_SIZE / 2);
+    __ ja(L_begin);
+
+    // needle_len is in elements, not bytes, for UTF-16
+    __ cmpq(needle_len, isUU ? OPT_NEEDLE_SIZE_MAX / 2 : OPT_NEEDLE_SIZE_MAX);
+    __ ja(L_begin);
+
+    // Handle short haystack and needle specially
+    // Generated code does not return - either found or not
+    highly_optimized_short_cases(ae, haystack, haystack_len, needle, needle_len, xmm0, xmm1, r8, r9, _masm);
+
+    // Set up jump tables.  Used when needle size <= NUMBER_OF_CASES
+    setup_jump_tables(ae, L_returnRBP, L_checkRangeAndReturn, L_bigCaseFixupAndReturn, &jump_table, &jump_table_1,
+                      _masm);
+
+    ////////////////////////////////////////////////////////////////////////////////////////
+    ////////////////////////////////////////////////////////////////////////////////////////
+    ////////////////////////////////////////////////////////////////////////////////////////
+
+    __ bind(L_begin);
```

Comment:
So far we have handled haystack <= 32 and needle_size <= 5 (?) in bytes. A high level algorithm description here is needed in comments to follow the code below.  A description of what are the various paths in terms of haystack and needle sizes and how to reason the assembly code below and make sure that all the paths are taken care of. Also the abstraction level suddenly changes here to detailed code below instead of methods for the various paths.

Reply:
I added a description.  Can you please check to ensure it meets your objective?  Thanks.